File: 문제풀이/온라인 저지/백준/# 12. Gold Ⅳ/# 11404. 플로이드/java/Main.java
import java.io.DataInputStream;
import java.io.DataOutputStream;

public class Main {

    public static void main(String[] args) {
        int n = i();
        int bus = i();
        int[][] cost = new int[n][n];
        int inf = 10_000_000;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cost[i][j] = i == j ? 0 : inf;
            }
        }

        for (int i = 0, a,b,c; i < bus; i++) {
            a = i() - 1;
            b = i() - 1;
            c = i();
            if (cost[a][b] > c) cost[a][b] = c;
        }

        for (int m = 0, alt; m < n; m++) {
            for (int s = 0; s < n; s++) {
                if ((s == m) || cost[s][m] == inf) continue;
                for (int e = 0; e < n; e++) {
                    if ((s == e) || (m == e) || cost[m][e] == inf) continue;
                    alt = cost[s][m] + cost[m][e];
                    if (cost[s][e] > alt) cost[s][e] = alt;
                }
            }
        }
        for (int[] arr : cost) {
            for (int c : arr) {
                writeInt((c == inf) ? 0 : c);
                writeByte(ASCII_space);
            }
            writeByte(ASCII_n);
        }
        flushBuffer();
    }

    private static final byte EOF = -1;
    private static final byte ASCII_n = 10;
    private static final byte ASCII_space = 32;
    private static final byte ASCII_minus = 45;
    private static final byte ASCII_0 = 48;
    private static final DataInputStream din = new DataInputStream(System.in);
    private static final DataOutputStream dout = new DataOutputStream(System.out);
    private static final byte[] inbuffer = new byte[65536];
    private static int inbufferpointer;
    private static int bytesread;
    private static final byte[] outbuffer = new byte[65536];
    private static int outbufferpointer;
    private static final byte[] bytebuffer = new byte[20];

    private static int i() {
        int v = 0;
        byte c = r();
        do {v = v * 10 + c - 48;} while (isDigit((c = r())));
        return v;
    }

    private static byte r() {
        if (inbufferpointer == bytesread) fillBuffer();
        return (bytesread == EOF) ? EOF : inbuffer[inbufferpointer++];
    }

    private static void fillBuffer() {
        try {
            bytesread = din.read(inbuffer, (inbufferpointer = 0), inbuffer.length);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static void writeInt(int i) {
        if (i == 0) {
            writeByte(ASCII_0);
        } else {
            if (i < 0) {
                writeByte(ASCII_minus);
                i = -i;
            }
            int index = 0;
            while (i > 0) {
                bytebuffer[index++] = (byte)(i % 10 + ASCII_0);
                i /= 10;
            }
            while (index-- > 0) {
                writeByte(bytebuffer[index]);
            }
        }
    }

    private static void writeByte(byte b) {
        if (outbufferpointer == outbuffer.length) flushBuffer();
        outbuffer[outbufferpointer++] = b;
    }

    private static void flushBuffer() {
        if (outbufferpointer != 0) {
            try {
                dout.write(outbuffer, 0, outbufferpointer);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
            outbufferpointer = 0;
        }
    }
    private static boolean isDigit(byte b) {
        return b >= ASCII_0;
    }
}

File: 문제풀이/온라인 저지/백준/# 12. Gold Ⅳ/# 11404. 플로이드/kotlin/Main.kt
fun main() {
    val n = i()
    var bus = i()
    val inf = 10_000_000
    val cost = Array(n) { i -> IntArray(n) { j -> if (i == j) 0 else inf } }
    var a: Int
    var b: Int
    var c: Int
    while (bus -- > 0) {
       a = i() - 1
       b = i() - 1
       c = i()
       if (cost[a][b] > c) cost[a][b] = c
    }
    var alt: Int
    for (m in 0 until n) {
        for (s in 0 until n) {
            if (s == m || cost[s][m] == inf) continue
            for (e in 0 until n) {
                if (s == e || m == e || cost[m][e] == inf) continue
                alt = cost[s][m] + cost[m][e]
                if (cost[s][e] > alt) cost[s][e] = alt
            }
        }
    }
    for (s in 0 until n) {
        for (e in 0 until n) {
            writeInt(cost[s][e].let { if (it == inf) 0 else it })
            writeByte(ASCII_space)
        }
        writeByte(ASCII_n)
    }
    flushBuffer()
}

private const val EOF = -1
private const val ASCII_n: Byte = 10
private const val ASCII_space: Byte = 32
private const val ASCII_minus: Byte = 45
private const val ASCII_0: Byte = 48
private val din = java.io.DataInputStream(System.`in`)
private val dout = java.io.DataOutputStream(System.out)
private val inbuffer = ByteArray(65536)
private var inbufferpointer = 0
private var bytesread = 0
private val outbuffer = ByteArray(65536)
private var outbufferpointer = 0
private val bytebuffer = ByteArray(20)

private fun i(): Int {
    var v = 0
    var c = r()
    do v = v * 10 + c - 48 while (isDigit(r().also { c = it }))
    return v
}

private fun r(): Byte {
    if (inbufferpointer == bytesread) fillBuffer()
    return if (bytesread == EOF) EOF.toByte() else inbuffer[inbufferpointer++]
}

private fun fillBuffer() {
    bytesread = din.read(inbuffer, 0.also { inbufferpointer = it }, inbuffer.size)
}

private fun writeInt(i: Int) {
    var i = i
    if (i == 0) {
        writeByte(ASCII_0)
    } else {
        if (i < 0) {
            writeByte(ASCII_minus)
            i = -i
        }
        var index = 0
        while (i > 0) {
            bytebuffer[index++] = (i % 10 + ASCII_0).toByte()
            i /= 10
        }
        while (index-- > 0) {
            writeByte(bytebuffer[index])
        }
    }
}

private fun writeByte(b: Byte) {
    if (outbufferpointer == outbuffer.size) flushBuffer()
    outbuffer[outbufferpointer++] = b
}

private fun flushBuffer() {
    if (outbufferpointer != 0) {
        dout.write(outbuffer, 0, outbufferpointer)
        outbufferpointer = 0
    }
}

private fun isDigit(b: Byte) = b >= ASCII_0